poseidon/poseidon.go

```diff
@@ -118,3 +121,56 @@ func Hash(inpBI []*big.Int) (*big.Int, error) {
 	rE.ToBigIntRegular(r)
 	return r, nil
 }
+
+// HashBytes returns a sponge hash of a msg byte slice split into blocks of 31 bytes
+func HashBytes(msg []byte) (*big.Int, error) {
+	// not used inputs default to zero
+	inputs := make([]*big.Int, spongeInputs)
+	for j := 0; j < spongeInputs; j++ {
+		inputs[j] = new(big.Int)
+	}
+	dirty := false
+	var hash *big.Int
+	var err error
+
+	k := 0
+	for i := 0; i < len(msg)/spongeChunkSize; i++ {
+		dirty = true
+		inputs[k].SetBytes(msg[spongeChunkSize*i : spongeChunkSize*(i+1)])
+		if k == spongeInputs-1 {
+			hash, err = Hash(inputs)
+			dirty = false
+			if err != nil {
+				return nil, err
+			}
+			inputs = make([]*big.Int, spongeInputs)
+			inputs[0] = hash
+			for j := 1; j < spongeInputs; j++ {
+				inputs[j] = new(big.Int)
+			}
+			k = 1
+		} else {
+			k++
+		}
+	}
+
+	if len(msg)%spongeChunkSize != 0 {
+		// the last chunk of the message is less than 31 bytes
+		// zero padding it, so that 0xdeadbeaf becomes
+		// 0xdeadbeaf000000000000000000000000000000000000000000000000000000
+		var buf [spongeChunkSize]byte
+		copy(buf[:], msg[(len(msg)/spongeChunkSize)*spongeChunkSize:])
+		inputs[k] = new(big.Int).SetBytes(buf[:])
+		dirty = true
+	}
+
+	if dirty {
+		// we haven't hashed something in the main sponge loop and need to do hash here
+		hash, err = Hash(inputs)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
There could be inputs not hashed in the main loop that need to be hashed later, e.g. we have msg length = 124, then inputs will look like this: `[chunk1, chunk2, chunk3, chunk4, 0, 0, ..., 0]` and `len(msg)%spongeChunkSize` would be 0, because 124/31 = 4 without the remaider

Reply:
if msg length = 125, then in the main loop inputs created will still be `[chunk1, chunk2, chunk3, chunk4, 0, 0, ..., 0]`, but there will also be 1 byte not included, which will be added to the inputs in `if len(msg)%spongeChunkSize != 0 {` block, to make `[chunk1, chunk2, chunk3, chunk4, chunk5, 0, ..., 0]`, where chunk5 will be zero-padded this 1 byte.